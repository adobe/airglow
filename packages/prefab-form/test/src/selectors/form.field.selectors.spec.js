/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import selector from '../../../src/selectors/form.field.selectors';

const state = {
  prefab: {
    size: {
      construct: {
        fieldType: 'number',
        required: true,
        max: 21,
        whenToValidate: 'blur',
        defaultValue: 3
      },
      store: {
        value: 29,
        hasBlurred: true
      }
    }
  }
};

describe('FormFieldSelectors', () => {
  let field;
  let dispatch;
  beforeEach(() => {
    field = selector('size');
    dispatch = sinon.spy();
  });
  describe('handlers', () => {
    let handlers;
    beforeEach(() => {
      handlers = field.handlers(dispatch);
    });
    it('fills out the bundle', () => {
      expect(field.state(state)).toMatchSnapshot();
    });
    it('triggers a change event', () => {
      handlers.onChange(32);
      expect(dispatch.getCall(0).args).toMatchSnapshot();
    });
    it('triggers a focus event', () => {
      handlers.onFocus();
      expect(dispatch.getCall(0).args).toMatchSnapshot();
    });
    it('triggers a blur event', () => {
      handlers.onBlur();
      expect(dispatch.getCall(0).args).toMatchSnapshot();
    });
    it('triggers a submit event', () => {
      handlers.onSubmit();
      expect(dispatch.getCall(0).args).toMatchSnapshot();
    });
    it('triggers a reset event', () => {
      handlers.onReset();
      expect(dispatch.getCall(0).args).toMatchSnapshot();
    });
  });
  it('creates a change action', () => {
    expect(field.changeAction(78)).toMatchSnapshot();
  });
  it('creates a submit action', () => {
    expect(field.submitAction()).toMatchSnapshot();
  });
  it('fetches the current value', () => {
    expect(field.value(state)).toBe(29);
  });
  it('fetches the current error', () => {
    expect(field.error(state)).toBe('error.number.max');
  });
});
