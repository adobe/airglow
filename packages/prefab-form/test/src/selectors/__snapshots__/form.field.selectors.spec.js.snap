// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldSelectors creates a change action 1`] = `
Object {
  "payload": Object {
    "construct": "size",
    "value": 78,
  },
  "type": "PREFAB_FORM_FIELD_VALUE",
}
`;

exports[`FormFieldSelectors creates a submit action 1`] = `
Object {
  "payload": Object {
    "construct": "size",
    "value": undefined,
  },
  "type": "PREFAB_FORM_FIELD_SUBMIT",
}
`;

exports[`FormFieldSelectors fetches the current state 1`] = `
Object {
  "dirty": true,
  "error": "error.number.max",
  "invalidMsg": "error.number.max",
  "required": true,
  "value": 29,
}
`;

exports[`FormFieldSelectors handlers fills out the bundle 1`] = `
Object {
  "dirty": true,
  "error": "error.number.max",
  "invalidMsg": "error.number.max",
  "required": true,
  "value": 29,
}
`;

exports[`FormFieldSelectors handlers triggers a blur event 1`] = `
Array [
  Object {
    "payload": Object {
      "construct": "size",
      "value": undefined,
    },
    "type": "PREFAB_FORM_FIELD_BLUR",
  },
]
`;

exports[`FormFieldSelectors handlers triggers a change event 1`] = `
Array [
  Object {
    "payload": Object {
      "construct": "size",
      "value": 32,
    },
    "type": "PREFAB_FORM_FIELD_VALUE",
  },
]
`;

exports[`FormFieldSelectors handlers triggers a focus event 1`] = `
Array [
  Object {
    "payload": Object {
      "construct": "size",
      "value": undefined,
    },
    "type": "PREFAB_FORM_FIELD_FOCUS",
  },
]
`;

exports[`FormFieldSelectors handlers triggers a reset event 1`] = `
Array [
  Object {
    "payload": Object {
      "construct": "size",
      "value": undefined,
    },
    "type": "PREFAB_FORM_FIELD_RESET",
  },
]
`;

exports[`FormFieldSelectors handlers triggers a submit event 1`] = `
Array [
  Object {
    "payload": Object {
      "construct": "size",
      "value": undefined,
    },
    "type": "PREFAB_FORM_FIELD_SUBMIT",
  },
]
`;
